### ISSUE TITLE
Bogus additional error compiling dynamic code in foreach in lambda for invalid code

### ISSUE THREAD

**Issue by hughbe** [dotnet/roslyn#19839]
**Version Used**: 
Microsoft (R) Visual C# Compiler version 2.1.0.61520

**Steps to Reproduce**:

Compile this code

```cs
using System.Collections.Generic;
using System.Dynamic;
using System.Linq;

namespace ConsoleApp1
{
    public class Program
    {
        public enum MappingType
        {
            Text,
            GeoPoint
        }
        
        public object Ingest()
        {
            return new ExpandoObject[0].Select(rawData =>
            {
                IDictionary<string, dynamic> mappedObject = new ExpandoObject();

                foreach (KeyValuePair<string, dynamic> keyValuePair in rawData)
                {
                    MappingType mapping = MappingType.GeoPoint;
                    object mappedValue = null;
                    switch (mapping)
                    {
                        case MappingType.Text:
                            mappedValue = keyValuePair.Value.ToString();
                            break;
                        case MappingType.GeoPoint:
                            mappedValue = keyValuePair
                            break;
                    }
                }

                return (ExpandoObject)mappedObject;
            }).ToList();
        }
    }
}
```

**Expected Behavior**:
Single error: 
```
Program.cs(31,55,31,55): error CS1002: ; expected
```

**Actual Behavior**:
Additional errors for valid code:
![untitled](https://cloud.githubusercontent.com/assets/1275900/26535417/7296f8b0-4457-11e7-8665-af4653ca861b.png)

Note that this additional error doesn't appear when compiling, only in the IDE
